extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java

```diff
@@ -81,29 +97,75 @@ public OpenCensusProtobufReader(
   }
 
   @Override
-  public CloseableIterator<InputRow> read()
+  public CloseableIterator<InputRow> read() throws IOException
   {
-    Supplier<Iterator<InputRow>> supplier = Suppliers.memoize(() -> readAsList().iterator());
-    return CloseableIterators.withEmptyBaggage(new Iterator<InputRow>() {
-      @Override
-      public boolean hasNext()
-      {
-        return supplier.get().hasNext();
-      }
-      @Override
-      public InputRow next()
-      {
-        return supplier.get().next();
+    return newReader(whichReader()).read();
+  }
+
+  class OpenCensusProtobufReaderInternal implements InputEntityReader
+  {
+    @Override
+    public CloseableIterator<InputRow> read()
+    {
+      return CloseableIterators.withEmptyBaggage(readAsList().iterator());
+    }
+
+    @Override
+    public CloseableIterator<InputRowListPlusRawValues> sample() throws IOException
+    {
+      return OpenCensusProtobufReader.this.sample();
+    }
+  }
+
+  public InputEntityReader newReader(ProtobufReader which)
+  {
+    switch(which) {
+      case OPEN_TELEMETRY:
+        return new OpenTelemetryMetricsProtobufReader(
+            dimensionsSpec,
+            source,
+            metricDimension,
+            valueDimension,
+            metricLabelPrefix,
+            resourceLabelPrefix
+        );
+      case OPEN_CENSUS:
+      default:
+        return new OpenCensusProtobufReaderInternal();
+    }
+  }
+
+  public ProtobufReader whichReader()
+  {
+    // assume InputEntity is always defined in a single classloader (the kafka-indexing-service classloader)
+    // so we only have to look it up once. To be completely correct we should cache the method based on classloader
+    MethodHandle getHeaderMethod = KafkaUtils.lookupGetHeaderMethod(
+        source.getEntity().getClass().getClassLoader(),
+        VERSION_HEADER_KEY
+    );
+
+    try {
+      byte[] versionHeader = (byte[]) getHeaderMethod.invoke(source.getEntity());
+      if (versionHeader != null) {
+        int version =
+            ByteBuffer.wrap(versionHeader).order(ByteOrder.LITTLE_ENDIAN).getInt();
+        if (version == OPENTELEMETRY_FORMAT_VERSION) {
+          return OPEN_TELEMETRY;
+        }
       }
-    });
+    }
+    catch (Throwable t) {
+      // assume input is opencensus if something went wrong
+    }
+    return OPEN_CENSUS;
   }
 
   List<InputRow> readAsList()
   {
     try {
-      return parseMetric(Metric.parseFrom(source.getBuffer()));
+      return parseMetric(Metric.parseFrom(source.open()));
```

Comment:
since we have a ByteEntity, we can avoid the overhead of parsing the underlying ByteBuffer via InputStream and read it directly like we did before using
```suggestion
      return parseMetric(Metric.parseFrom(source.getEntity().getBuffer()));
```

Reply:
looking a bit at the protobuf code, it looks like it has some pretty sophisticated optimizations for ByteBuffers and will read the underlying array directly or bypass the JVM bounds check using unsafe if possible to get the best performance.
It is likely that this causes the ByteBuffer position to not advance and trigger the check here https://github.com/apache/druid/blob/6346b9561df7a92557acb51a56585084a0eb8633/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SettableByteEntity.java#L97-L99

A simple workaround might be to call `buffer.position(buffer.position() + buffer.remaining())` when done parsing from the buffer.

While the additional check is nice, I think it breaks the existing "contract" in that previously there was no requirement to read out the buffer entirely,. I believe it was mainly added to avoid some kind of unexpected bug or concurrency issue with SettableByteEntity

Reply:
I don't think we have performance issues by calling `open` here. If we look at the code of [SettableByteEntity#open](https://github.com/confluentinc/druid/blob/e0c3831aed8f8b904719d8c14b18284e5c72f3f9/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SettableByteEntity.java#LL80C9-L80C9) all it does is to check if the input stream is already open and if it doesn't it returns it. (I want to say I checked that before I write the call here but it's been a few days).

Looking there at [SettableByteEntity#setEntity](https://github.com/confluentinc/druid/blob/major-druid-upgrade-confluent-rc2/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SettableByteEntity.java#L56-L61) which I assume is called every time before a call to `open` in our case, it seems that it does something similar to what you describe. It sets the buffer for that inputstream and resets the `opened` flag. So, in a sense it re-uses the bytebuffer that is given via the `entity` that is passed. 

So not a traditional `open` with its costs it seems (but I still might be missing something given my low familiarity with this code). Wdyt?

Reply:
I think it's still more efficient for the protobuf decoder to have access to the buffer, since it can skip ahead rather than having to consume all the bytes in a linear fashion from the inputstream. The InputStream will also incur the cost of bounds checks on the ByteBuffer, which the protobuf decoder currently goes out of its way to avoid.

Reply:
the fact that SettableByteEntity seems geared towards returning an InputStream is something we'll probably want to improve upstream. Part of the reason we have ByteEntity is so that we can optimize access to the underlying buffer for performance. I don't think anything stands in our way here to get the underlying ByteBuffer now, and we can improve the API to make that easier in the future.

Reply:
I see what you mean now. I didn't go into that depth with the protobuf calls (still a bit challenging to get to the non-sequential parsing you are referring to). 

The main other example with protobuf beyond the classes we discuss here that I can find is in 
[ProtobufReader#intermediateRowIterator](https://github.com/confluentinc/druid/blob/e0c3831aed8f8b904719d8c14b18284e5c72f3f9/extensions-core/protobuf-extensions/src/main/java/org/apache/druid/data/input/protobuf/ProtobufReader.java#LL83)
and while it passes a ByteBuffer it first reads and converts the stream into a bytearray. 

```
  Iterators.singletonIterator(protobufBytesDecoder.parse(ByteBuffer.wrap(IOUtils.toByteArray(source.open()))))
```

Reply:
Changed the code to set the buffer position explicitly to the limit. Let me know if the code now matches what you had in mind.